File: src/stewart_plugin/ik.cpp
#include <cmath>
#include "eigen3/Eigen/Core"
#include "std_msgs/Float32MultiArray.h"


Eigen::Matrix<float, 4, 4> transformation_matrix(float x, float y, float z, float r, float p, float yaw)
{
    Eigen::Matrix<float, 4, 4> T;
    T << cos(yaw)*cos(p), -sin(yaw)*cos(r) + cos(yaw)*sin(p)*sin(r),  sin(yaw)*sin(r)+cos(yaw)*sin(p)*cos(r), x,
         sin(yaw)*cos(p),  cos(yaw)*cos(r) + sin(yaw)*sin(p)*sin(r), -cos(yaw)*sin(r)+sin(yaw)*sin(p)*cos(r), y,
                 -sin(p),                             cos(p)*sin(r),                         cos(p)*cos(yaw), z,
                       0,                                         0,                                       0, 1;
    return T;
}


Eigen::Matrix<float, 6, 1> ik(const std_msgs::Float32MultiArray::ConstPtr& goalPos)
{
        Eigen::Matrix<float, 1, 6> goalJointPos = {};
        goalJointPos << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;

        float height = 2.0;
        Eigen::Matrix<float, 6, 4> b, p;


        b << -0.101,    0.8, 0.25, 1,
              0.101,    0.8, 0.25, 1,
              0.743, -0.313, 0.25, 1,
              0.642, -0.487, 0.25, 1,
             -0.643, -0.486, 0.25, 1,
             -0.744, -0.311, 0.25, 1;

        p << -0.642,  0.487, -0.05, 1,
              0.642,  0.487, -0.05, 1,
              0.743,  0.313, -0.05, 1,
              0.101,   -0.8, -0.05, 1,
             -0.101,   -0.8, -0.05, 1,
             -0.743,  0.313, -0.05, 1;


        float x = goalPos->data[0];
        float y = goalPos->data[1];
        float z = goalPos->data[2];
        float roll = goalPos->data[3];
        float pitch = goalPos->data[4];
        float yaw = goalPos->data[5];
        Eigen::Matrix<float, 4, 4> T = transformation_matrix(x, y, z, roll, pitch, yaw);
        for (size_t i = 0; i < 6; i++)
        {
            Eigen::Matrix<float, 4, 1> length = T*p.row(i).transpose() - b.row(i).transpose();
            goalJointPos[i] = sqrt(pow(length(0), 2) + pow(length(1), 2) + pow(length(2), 2)) - height;
        }

        return goalJointPos;
}

File: src/stewart/scripts/ik.py
import numpy as np
import os
import sys

def transformation_matrix(x, y, z, roll, pitch, yaw):
    """
    Transformation matrix
    INPUT: SCALARS
    OUTPUT: 4X4 NUMPY ARRAY
    """
    T = np.array([[np.cos(yaw) * np.cos(pitch), -np.sin(yaw)*np.cos(roll) + np.cos(yaw)*np.sin(pitch)*np.sin(roll),  np.sin(yaw)*np.sin(roll)+np.cos(yaw)*np.sin(pitch)*np.cos(roll), x],
                  [np.sin(yaw)*np.cos(pitch),  np.cos(yaw)*np.cos(roll) + np.sin(yaw)*np.sin(pitch)*np.sin(roll), -np.cos(yaw)*np.sin(roll)+np.sin(yaw)*np.sin(pitch)*np.cos(roll), y],
                  [-np.sin(pitch),  np.cos(pitch)*np.sin(roll), np.cos(pitch)*np.cos(yaw), z],
                  [0., 0., 0., 1.]])
    return T

def ikin(goalPos):
    """
    Ikin method
    INPUT: goalPos as numpy column pls!!!!!
    OUTPUT: List
    """
    # All inputs as numpy array NO ROS MESSAGEs
    base_length = 2

    b = np.array([[-0.101,    0.8, 0.25, 1],
                  [0.101,    0.8, 0.25, 1],
                  [0.743, -0.313, 0.25, 1],
                  [0.642, -0.487, 0.25, 1],
                  [-0.643, -0.486, 0.25, 1],
                  [-0.744, -0.311, 0.25, 1]])

    p = np.array([[-0.642,  0.487, -0.05, 1],
                  [0.642,  0.487, -0.05, 1],
                  [0.743,  0.313, -0.05, 1],
                  [0.101,   -0.8, -0.05, 1],
                  [-0.101,   -0.8, -0.05, 1],
                  [-0.743,  0.313, -0.05, 1]])


    # Acquiring from numpy input array
    # Goal position as a column vector pls
    x, y, z, roll, pitch, yaw = goalPos[0][0], goalPos[1][0], goalPos[2][0], goalPos[3][0], goalPos[4][0], goalPos[5][0]

    T = transformation_matrix(x, y, z, roll, pitch, yaw)
    goal_joint_pos = []
    for i in range(6):
        length = T @ p[i, :].reshape((-1, 1)) - b[i, :].reshape((-1, 1))
        joint_pos = np.sqrt(np.sum(length[:3, :]**2)) - base_length
        goal_joint_pos.append(joint_pos)
    return goal_joint_pos
